### PR TITLE
V11 migration

### PR DESCRIPTION
**Pull request by Alchez**

### CODE REVIEW

File: awesome_cart/zscript.py [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+  
+  arg_values = []
+  for arg in args[1:]:
+    if isinstance(arg, basestring):
+      arg_values += [arg.format(**context)]
+    elif isinstance(arg, list) or isinstance(arg, tuple):
+      arg_values += [execCommand(arg, context, depth+1)]
+    else:
+      arg_values += [arg]
+
```
Comment (vjFaLk) [inline awesome_cart/zscript.py]: Codacy Issue found: [Trailing whitespace](https://app.codacy.com/manual/DigiThinkIT/awesome_cart/pullRequest?prid=4737459)

File: awesome_cart/awc.py [hunk 2]
```diff
@@ -1652,6 +1678,133 @@ def cart(data=None, action=None):
...
+
+		if remove_success:
+			# replace items with filtered version
+			awc["items"] = awc_items
+			removed_ids += list(set(removed_quotation_item_ids) - set(removed_ids))
+
+	if len(removed_ids) > 0:
+		quotation_items = [ \
+			itm for itm in quotation.get("items", []) \
+									if itm.name not in removed_ids \
```
Comment (vjFaLk) [inline awesome_cart/awc.py]: Codacy Issue found: [continuation line unaligned for hanging indent](https://app.codacy.com/manual/DigiThinkIT/awesome_cart/pullRequest?prid=4737459)

File: awesome_cart/zscript.py [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+# arguments passed to the command.
+#
+# Example:
+#   [ "and", true, false ]
+# This will evaluate to false and is equivalent to all([true, false])
+# See the COMMANDS map below
+#
+# Example:
+#   [ 
+#     "==",
```
Comment (vjFaLk) [inline awesome_cart/zscript.py]: Codacy Issue found: [Trailing whitespace](https://app.codacy.com/manual/DigiThinkIT/awesome_cart/pullRequest?prid=4737459)

File: awesome_cart/awc.py [hunk 4]
```diff
@@ -1537,6 +1595,9 @@ def cart(data=None, action=None):
 		removed_ids = []
 
 		success, removed_ids, awc_items = remove_from_cart(data, awc["items"])
+		if quotation and quotation.coupon_code:
+			coupon_success, coupon_msg, quotation_is_dirty, coupon_is_valid, coupon_removed_ids = apply_coupon(awc, awc_session, quotation, quotation.coupon_code, quotation_is_dirty)
```
Comment (vjFaLk) [inline awesome_cart/awc.py]: Codacy Issue found: [line too long (173 > 159 characters)](https://app.codacy.com/manual/DigiThinkIT/awesome_cart/pullRequest?prid=4737459)

File: awesome_cart/awc.py [hunk 5]
```diff
@@ -1652,6 +1678,133 @@ def cart(data=None, action=None):
...
+		quotation_items = [ \
+			itm for itm in quotation.get("items", []) \
+									if itm.name not in removed_ids \
+		]
+		awc["items"] = [ i for i in awc.get("items", []) if i.get("id") not in removed_ids ]
+		quotation.set("items", quotation_items)
+
+	return (success, msg, quotation_is_dirty, is_valid, removed_ids)
+
+def find_item_by_insert_id(awc_items, id):
```
Comment (vjFaLk) [inline awesome_cart/awc.py]: Codacy Issue found: [Redefining built-in 'id'](https://app.codacy.com/manual/DigiThinkIT/awesome_cart/pullRequest?prid=4737459)

File: awesome_cart/awc.py [hunk 6]
```diff
@@ -4,6 +4,7 @@
 import traceback
 import frappe
 import datetime
+import json
```
Comment (vjFaLk) [inline awesome_cart/awc.py]: Codacy Issue found: [Reimport 'json' (imported line 3)](https://app.codacy.com/manual/DigiThinkIT/awesome_cart/pullRequest?prid=4737459)

File: awesome_cart/awesome_cart/doctype/awc_coupon/awc_coupon.py [hunk 7]
```diff
@@ -348,12 +349,62 @@ def is_coupon_valid(coupon_code, customer, now=None):
...
+		elif insert_item.get("insert_logic") == "Maximum":
+			insert_logic = [
+				"and",
+				["==", "{coupon_code}", coupon_code],
+				["<=", ["float", "{order_total}"], insert_item.get("insert_max")]
+			]
+
+		if insert_item.get("stock_limit"):
+			insert_logic += [
+				">",
```
Comment (vjFaLk) [inline awesome_cart/awesome_cart/doctype/awc_coupon/awc_coupon.py]: Codacy Issue found: [Trailing whitespace](https://app.codacy.com/manual/DigiThinkIT/awesome_cart/pullRequest?prid=4737459)

File: awesome_cart/awc.py [hunk 8]
```diff
@@ -1512,6 +1562,10 @@ def cart(data=None, action=None):
 
 		if quotation:
 			quotation_is_dirty=True
+			if quotation.coupon_code:
+				coupon_success, coupon_msg, quotation_is_dirty, coupon_is_valid, coupon_removed_ids = apply_coupon(awc, awc_session, quotation, quotation.coupon_code, quotation_is_dirty)
```
Comment (vjFaLk) [inline awesome_cart/awc.py]: Codacy Issue found: [Unused variable 'coupon_is_valid'](https://app.codacy.com/manual/DigiThinkIT/awesome_cart/pullRequest?prid=4737459)

File: awesome_cart/awesome_cart/doctype/awc_coupon/awc_coupon.py [hunk 9]
```diff
@@ -348,12 +349,62 @@ def is_coupon_valid(coupon_code, customer, now=None):
...
-			"lock_qty": insert_item.get("lock_qty", False),
-			"total_is_greater_than": insert_item.get("total_is_greater_than", 0)
-		})
+			"lock_qty": insert_item.get("lock_qty", False)
+		}
+
+		# Defines a custom insert logic to later test and remove item depending on this logic result.
+		if insert_item.get("insert_logic") == "Between":
+			insert_logic = [
+				"and",
```
Comment (vjFaLk) [inline awesome_cart/awesome_cart/doctype/awc_coupon/awc_coupon.py]: Codacy Issue found: [Trailing whitespace](https://app.codacy.com/manual/DigiThinkIT/awesome_cart/pullRequest?prid=4737459)

File: awesome_cart/awc.py [hunk 10]
```diff
@@ -1569,50 +1630,11 @@ def cart(data=None, action=None):
...
-					is_valid = False
-					success = False
-					awc["discounts"] = None
-					if not has_total_limit:
-						msg = "Coupon is invalid for the current cart."
-				else:
-					awc["discounts"] = coupon_state
-
-					insert_items = frappe.response.get("coupon_insert_items", False)
+			success, msg, quotation_is_dirty, is_valid, removed_ids = apply_coupon(awc, awc_session, quotation, coupon, quotation_is_dirty)
```
Comment (vjFaLk) [inline awesome_cart/awc.py]: Codacy Issue found: [Unused variable 'is_valid'](https://app.codacy.com/manual/DigiThinkIT/awesome_cart/pullRequest?prid=4737459)

File: awesome_cart/zscript.py [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+# The first value in the array is always the command, all other values are the
+# arguments passed to the command.
+#
+# Example:
+#   [ "and", true, false ]
+# This will evaluate to false and is equivalent to all([true, false])
+# See the COMMANDS map below
+#
+# Example:
+#   [
```
Comment (vjFaLk) [inline awesome_cart/zscript.py]: Codacy Issue found: [Trailing whitespace](https://app.codacy.com/manual/DigiThinkIT/awesome_cart/pullRequest?prid=4737459)

File: awesome_cart/awc.py [hunk 12]
```diff
@@ -1652,6 +1678,133 @@ def cart(data=None, action=None):
 			"message": "Unknown Command."
 		}, awc_session, quotation)
 
+def apply_coupon(awc, awc_session, quotation, coupon, quotation_is_dirty):
```
Comment (vjFaLk) [inline awesome_cart/awc.py]: Codacy Issue found: [apply_coupon is too complex (17)](https://app.codacy.com/manual/DigiThinkIT/awesome_cart/pullRequest?prid=4737459)

File: awesome_cart/zscript.py [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+  if isinstance(script, basestring):
+    script_obj = json.loads(script)
+
+  if not isinstance(script_obj, list):
+    raise Exception("ZScripts must be array of arrays")
+
+  return execCommand(script_obj, context)
+
+def execCommand(args, context, depth=0):
+  pad = " " * depth * 2
```
Comment (vjFaLk) [inline awesome_cart/zscript.py]: Codacy Issue found: [Unused variable 'pad'](https://app.codacy.com/manual/DigiThinkIT/awesome_cart/pullRequest?prid=4737459)

File: awesome_cart/awc.py (hunk 6 above)
Comment (vjFaLk) [inline awesome_cart/awc.py]: Codacy Issue found: [redefinition of unused 'json' from line 3](https://app.codacy.com/manual/DigiThinkIT/awesome_cart/pullRequest?prid=4737459)

File: awesome_cart/awc.py [hunk 14]
```diff
@@ -1,24 +1,25 @@
...
 import json
 import traceback
-import frappe
-import datetime
 import uuid
 
-from frappe import _dict, _
-from frappe.utils import cint, cstr, random_string, flt, get_datetime
+import frappe
+from dti_devtools.debug import log, pretty_json
```
Comment (vjFaLk) [inline awesome_cart/awc.py]: Codacy Issue found: [Unused pretty_json imported from dti_devtools.debug](https://app.codacy.com/manual/DigiThinkIT/awesome_cart/pullRequest?prid=4737459)

File: awesome_cart/awc.py [hunk 15]
```diff
@@ -1512,6 +1561,10 @@ def cart(data=None, action=None):
 
 		if quotation:
 			quotation_is_dirty=True
+			if quotation.coupon_code:
+				coupon_success, coupon_msg, quotation_is_dirty, coupon_is_valid, coupon_removed_ids = apply_coupon(awc, awc_session, quotation, quotation.coupon_code, quotation_is_dirty)
```
Comment (vjFaLk) [inline awesome_cart/awc.py]: Codacy Issue found: [Unused variable 'coupon_is_valid'](https://app.codacy.com/manual/DigiThinkIT/awesome_cart/pullRequest?prid=4737459)

File: awesome_cart/awc.py [hunk 16]
```diff
@@ -1652,6 +1677,133 @@ def cart(data=None, action=None):
 			"message": "Unknown Command."
 		}, awc_session, quotation)
 
+def apply_coupon(awc, awc_session, quotation, coupon, quotation_is_dirty):
```
Comment (vjFaLk) [inline awesome_cart/awc.py]: Codacy Issue found: [apply_coupon is too complex (17)](https://app.codacy.com/manual/DigiThinkIT/awesome_cart/pullRequest?prid=4737459)

File: awesome_cart/zscript.py [hunk 17]
```diff
@@ -0,0 +1,96 @@
...
```
Comment (vjFaLk) [inline awesome_cart/zscript.py]: Codacy Issue found: [local variable 'pad' is assigned to but never used](https://app.codacy.com/manual/DigiThinkIT/awesome_cart/pullRequest?prid=4737459)

Comment (vjFaLk) [inline awesome_cart/zscript.py]: Codacy Issue found: [Unused variable 'pad'](https://app.codacy.com/manual/DigiThinkIT/awesome_cart/pullRequest?prid=4737459)

File: awesome_cart/awc.py [hunk 18]
```diff
@@ -1,24 +1,25 @@
...
-import datetime
 import uuid
 
-from frappe import _dict, _
-from frappe.utils import cint, cstr, random_string, flt, get_datetime
+import frappe
+from dti_devtools.debug import log, pretty_json
 from erpnext.stock.get_item_details import apply_price_list_on_item
+from frappe import _, _dict
+from frappe.utils import cint, cstr, flt, get_datetime, random_string
```
Comment (vjFaLk) [inline awesome_cart/awc.py]: Codacy Issue found: ['frappe.utils.random_string' imported but unused](https://app.codacy.com/manual/DigiThinkIT/awesome_cart/pullRequest?prid=4737459)

File: awesome_cart/awc.py [hunk 19]
```diff
@@ -1537,6 +1594,9 @@ def cart(data=None, action=None):
 		removed_ids = []
 
 		success, removed_ids, awc_items = remove_from_cart(data, awc["items"])
+		if quotation and quotation.coupon_code:
+			coupon_success, coupon_msg, quotation_is_dirty, coupon_is_valid, coupon_removed_ids = apply_coupon(awc, awc_session, quotation, quotation.coupon_code, quotation_is_dirty)
```
Comment (vjFaLk) [inline awesome_cart/awc.py]: Codacy Issue found: [line too long (173 > 159 characters)](https://app.codacy.com/manual/DigiThinkIT/awesome_cart/pullRequest?prid=4737459)

File: awesome_cart/awc.py [hunk 20]
```diff
@@ -1652,6 +1677,133 @@ def cart(data=None, action=None):
...
+		quotation_items = [ \
+			itm for itm in quotation.get("items", []) \
+									if itm.name not in removed_ids \
+		]
+		awc["items"] = [ i for i in awc.get("items", []) if i.get("id") not in removed_ids ]
+		quotation.set("items", quotation_items)
+
+	return (success, msg, quotation_is_dirty, is_valid, removed_ids)
+
+def find_item_by_insert_id(awc_items, id):
```
Comment (vjFaLk) [inline awesome_cart/awc.py]: Codacy Issue found: [Redefining built-in 'id'](https://app.codacy.com/manual/DigiThinkIT/awesome_cart/pullRequest?prid=4737459)

File: awesome_cart/awc.py (hunk 18 above)
Comment (vjFaLk) [inline awesome_cart/awc.py]: Codacy Issue found: [Unused random_string imported from frappe.utils](https://app.codacy.com/manual/DigiThinkIT/awesome_cart/pullRequest?prid=4737459)

File: awesome_cart/awc.py [hunk 21]
```diff
@@ -1569,50 +1629,11 @@ def cart(data=None, action=None):
...
-					awc["discounts"] = None
-					if not has_total_limit:
-						msg = "Coupon is invalid for the current cart."
-				else:
-					awc["discounts"] = coupon_state
-
-					insert_items = frappe.response.get("coupon_insert_items", False)
-
-					if insert_items:
+			success, msg, quotation_is_dirty, is_valid, removed_ids = apply_coupon(awc, awc_session, quotation, coupon, quotation_is_dirty)
```
Comment (vjFaLk) [inline awesome_cart/awc.py]: Codacy Issue found: [Unused variable 'is_valid'](https://app.codacy.com/manual/DigiThinkIT/awesome_cart/pullRequest?prid=4737459)

File: awesome_cart/awc.py [hunk 22]
```diff
@@ -1652,6 +1677,133 @@ def cart(data=None, action=None):
...
+
+		if remove_success:
+			# replace items with filtered version
+			awc["items"] = awc_items
+			removed_ids += list(set(removed_quotation_item_ids) - set(removed_ids))
+
+	if len(removed_ids) > 0:
+		quotation_items = [ \
+			itm for itm in quotation.get("items", []) \
+									if itm.name not in removed_ids \
```
Comment (vjFaLk) [inline awesome_cart/awc.py]: Codacy Issue found: [continuation line unaligned for hanging indent](https://app.codacy.com/manual/DigiThinkIT/awesome_cart/pullRequest?prid=4737459)

File: awesome_cart/awc.py (hunk 14 above)
Comment (vjFaLk) [inline awesome_cart/awc.py]: Codacy Issue found: ['dti_devtools.debug.pretty_json' imported but unused](https://app.codacy.com/manual/DigiThinkIT/awesome_cart/pullRequest?prid=4737459)

Comment (vjFaLk) [inline awesome_cart/awc.py]: Codacy Issue found: [Unused pretty_json imported from dti_devtools.debug](https://app.codacy.com/manual/DigiThinkIT/awesome_cart/pullRequest?prid=4737459)

File: awesome_cart/zscript.py (hunk 17 above)
Comment (vjFaLk) [inline awesome_cart/zscript.py]: Codacy Issue found: [local variable 'pad' is assigned to but never used](https://app.codacy.com/manual/DigiThinkIT/awesome_cart/pullRequest?prid=4737459)

File: awesome_cart/awc.py (hunk 14 above)
Comment (vjFaLk) [inline awesome_cart/awc.py]: Codacy Issue found: ['dti_devtools.debug.pretty_json' imported but unused](https://app.codacy.com/manual/DigiThinkIT/awesome_cart/pullRequest?prid=4737459)

File: awesome_cart/zscript.py (hunk 17 above)
Comment (vjFaLk) [inline awesome_cart/zscript.py]: Codacy Issue found: [Unused variable 'pad'](https://app.codacy.com/manual/DigiThinkIT/awesome_cart/pullRequest?prid=4737459)

File: awesome_cart/awc.py (hunk 18 above)
Comment (vjFaLk) [inline awesome_cart/awc.py]: Codacy Issue found: [Unused random_string imported from frappe.utils](https://app.codacy.com/manual/DigiThinkIT/awesome_cart/pullRequest?prid=4737459)

File: awesome_cart/awc.py (hunk 21 above)
Comment (vjFaLk) [inline awesome_cart/awc.py]: Codacy Issue found: [Unused variable 'is_valid'](https://app.codacy.com/manual/DigiThinkIT/awesome_cart/pullRequest?prid=4737459)

File: awesome_cart/awc.py (hunk 16 above)
Comment (vjFaLk) [inline awesome_cart/awc.py]: Codacy Issue found: [apply_coupon is too complex (17)](https://app.codacy.com/manual/DigiThinkIT/awesome_cart/pullRequest?prid=4737459)

File: awesome_cart/awc.py (hunk 20 above)
Comment (vjFaLk) [inline awesome_cart/awc.py]: Codacy Issue found: [Redefining built-in 'id'](https://app.codacy.com/manual/DigiThinkIT/awesome_cart/pullRequest?prid=4737459)

File: awesome_cart/awc.py (hunk 19 above)
Comment (vjFaLk) [inline awesome_cart/awc.py]: Codacy Issue found: [line too long (173 > 159 characters)](https://app.codacy.com/manual/DigiThinkIT/awesome_cart/pullRequest?prid=4737459)

File: awesome_cart/awc.py (hunk 18 above)
Comment (vjFaLk) [inline awesome_cart/awc.py]: Codacy Issue found: ['frappe.utils.random_string' imported but unused](https://app.codacy.com/manual/DigiThinkIT/awesome_cart/pullRequest?prid=4737459)

File: awesome_cart/awc.py (hunk 15 above)
Comment (vjFaLk) [inline awesome_cart/awc.py]: Codacy Issue found: [Unused variable 'coupon_is_valid'](https://app.codacy.com/manual/DigiThinkIT/awesome_cart/pullRequest?prid=4737459)

File: awesome_cart/awc.py (hunk 22 above)
Comment (vjFaLk) [inline awesome_cart/awc.py]: Codacy Issue found: [continuation line unaligned for hanging indent](https://app.codacy.com/manual/DigiThinkIT/awesome_cart/pullRequest?prid=4737459)

File: awesome_cart/awc.py [hunk 23]
```diff
@@ -1652,6 +1676,133 @@ def cart(data=None, action=None):
...
```
Comment (vjFaLk) [inline awesome_cart/awc.py:1783]: Codacy Issue found: [continuation line unaligned for hanging indent](https://app.codacy.com/manual/DigiThinkIT/awesome_cart/pullRequest?prid=4737459)

File: awesome_cart/awc.py [hunk 24]
```diff
@@ -1652,6 +1676,133 @@ def cart(data=None, action=None):
 			"message": "Unknown Command."
 		}, awc_session, quotation)
 
+def apply_coupon(awc, awc_session, quotation, coupon, quotation_is_dirty):
```
Comment (vjFaLk) [inline awesome_cart/awc.py:1679]: Codacy Issue found: [apply_coupon is too complex (17)](https://app.codacy.com/manual/DigiThinkIT/awesome_cart/pullRequest?prid=4737459)

File: awesome_cart/awc.py [hunk 25]
```diff
@@ -1569,50 +1628,11 @@ def cart(data=None, action=None):
...
-					awc["discounts"] = None
-					if not has_total_limit:
-						msg = "Coupon is invalid for the current cart."
-				else:
-					awc["discounts"] = coupon_state
-
-					insert_items = frappe.response.get("coupon_insert_items", False)
-
-					if insert_items:
+			success, msg, quotation_is_dirty, is_valid, removed_ids = apply_coupon(awc, awc_session, quotation, coupon, quotation_is_dirty)
```
Comment (vjFaLk) [inline awesome_cart/awc.py:1631]: Codacy Issue found: [Unused variable 'is_valid'](https://app.codacy.com/manual/DigiThinkIT/awesome_cart/pullRequest?prid=4737459)

File: awesome_cart/awc.py [hunk 26]
```diff
@@ -1537,6 +1593,9 @@ def cart(data=None, action=None):
 		removed_ids = []
 
 		success, removed_ids, awc_items = remove_from_cart(data, awc["items"])
+		if quotation and quotation.coupon_code:
+			coupon_success, coupon_msg, quotation_is_dirty, coupon_is_valid, coupon_removed_ids = apply_coupon(awc, awc_session, quotation, quotation.coupon_code, quotation_is_dirty)
```
Comment (vjFaLk) [inline awesome_cart/awc.py:1597]: Codacy Issue found: [line too long (173 > 159 characters)](https://app.codacy.com/manual/DigiThinkIT/awesome_cart/pullRequest?prid=4737459)

File: awesome_cart/awc.py [hunk 27]
```diff
@@ -1512,6 +1560,10 @@ def cart(data=None, action=None):
 
 		if quotation:
 			quotation_is_dirty=True
+			if quotation.coupon_code:
+				coupon_success, coupon_msg, quotation_is_dirty, coupon_is_valid, coupon_removed_ids = apply_coupon(awc, awc_session, quotation, quotation.coupon_code, quotation_is_dirty)
```
Comment (vjFaLk) [inline awesome_cart/awc.py:1564]: Codacy Issue found: [Unused variable 'coupon_is_valid'](https://app.codacy.com/manual/DigiThinkIT/awesome_cart/pullRequest?prid=4737459)

File: awesome_cart/zscript.py [hunk 28]
```diff
@@ -0,0 +1,96 @@
...
+  if isinstance(script, basestring):
+    script_obj = json.loads(script)
+
+  if not isinstance(script_obj, list):
+    raise Exception("ZScripts must be array of arrays")
+
+  return execCommand(script_obj, context)
+
+def execCommand(args, context, depth=0):
+  # pad = " " * depth * 2
```
Comment (vjFaLk) [inline awesome_cart/zscript.py:84]: Codacy Issue found: [indentation is not a multiple of four (comment)](https://app.codacy.com/manual/DigiThinkIT/awesome_cart/pullRequest?prid=4737459)

**Review by DeveloperMetal — APPROVED** [review]

**Pull request merged**